# QC reports from VCF files

Set of R markdown to generate an html QC report for small variants and structural variants VCF.

## How to generate the reports

1. Put the input files in the same folder with the .Rmd files
2. Generate HTML reports using

```
R -e "rmarkdown::render('QC_report_SV.Rmd', output_file='report_SV.html')"
R -e "rmarkdown::render('QC_report_smallvars.Rmd', output_file='report_smallvars.html')"
```

## Input files
By default the report expects the files described below. The following parameters can be set at commandline to change title and input files

```
params:
  doc_title: "Small variants QC"
  vcf_stats: "VCF.stats"
  chr_counts: "VCF_varsPerChrom.stats"
  ancestry: "ancestry.tsv"
```

So for example you can use

```
R -e "rmarkdown::render('QC_report_smallvars.Rmd', output_file='report_smallvars.html',params=list(doc_title="mytitle", vcf_stats="my_vcfstats.txt:, ...)"
```

### Input for small variants
By default the report expects the following files:
- **VCF.stats**: stats generated using `bcftools stats -s-`
- **VCF_varsPerChrom.stats**: count of variants per chromosome generated using `bcftools index --stats`
- **ancestry.tsv** file generated by `somalier ancestry`

### Input for SV variants
- **SV_per_sample_info.tsv**: Per sample data generated with

```
bcftools query -i 'FILTER == "PASS"' -f '[%CHROM\t%POS\t%SVTYPE\t%SVLEN\t%AF\t%SAMPLE\t%GT\t%SQ\t%DP\t%CN\n]' input_vcf.vcf.gz >
```

- **SV_per_variant_info.tsv**: Per variant data generated with

```
bcftools query -i 'FILTER == "PASS"' -H -f '%CHROM\t%POS\t%ID\t%FILTER\t%SVTYPE\t%SVLEN\t%CIPOS95\t%AF\n' input_vcf.vcf.gz > SV_
```

## Dependencies
- ggplot2
- dplyr
- tidyr
- knitr
- patchwork
